### PR TITLE
used MaxUnlockingChunks from config

### LINKED ISSUE #12227: `MaxUnlockingChunks` should be configurable via Runtime
StakingLedger in the `pallet-staking` is using a hardcoded value of [MaxUnlockingChunks](https://github.com/paritytech/substrate/blob/master/frame/staking/src/lib.rs#L353). Pallet-staking also has a [MaxUnlockingChunks in the configuration](https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/mod.rs#L192) which is not used anywhere.

I discussed this with @kianenigma, we should remove the hard coded value and use the value we get from Config.

### PR DESCRIPTION
**Pull request by MrishoLukamba**
@kianenigma @ggwpez 
Fixes #12227

### CODE REVIEW

**Comment by MrishoLukamba** [comment]
Solved issue #12227

**Comment by kianenigma** [comment]
Please follow this syntax to close the linked issue https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword.

**Review by kianenigma — COMMENTED** [review]
You should document in the config type parameter that if this value is ever set, and then decreased, it leads to a corrupt state and a migration is needed. We also need a test to demonstrate this. 

@Ank4n please note that we have the same story with `MaxNominations` as well.

**Comment by Ank4n** [comment]
@MrishoLukamba Can you also add a test that documents what happens in the scenario where MaxUnlockingChunks is reduced without migration?
Here are couple of examples you can refer
https://github.com/paritytech/substrate/blob/ankan/bound-staking-storage-trivial/frame/staking/src/tests.rs#L4872
https://github.com/paritytech/substrate/blob/ankan/bound-staking-storage-trivial/frame/staking/src/tests.rs#L5367

**Comment by MrishoLukamba** [comment]
> @MrishoLukamba Can you also add a test that documents what happens in the scenario where MaxUnlockingChunks is reduced without migration? Here are couple of examples you can refer https://github.com/paritytech/substrate/blob/ankan/bound-staking-storage-trivial/frame/staking/src/tests.rs#L4872 https://github.com/paritytech/substrate/blob/ankan/bound-staking-storage-trivial/frame/staking/src/tests.rs#L5367

Okey let me get into it, if anything I will ask

**Comment by MrishoLukamba** [comment]
@Ank4n @kianenigma There is an issue which I noticed, StakingLedger.staking  BoundedVec does not push new object (UnlockingChunks) instead it just increments the value inside UnlockingChunks. So the MaxUnlockingChunks bound in boundedVec is not quite useful, 
I think is a bug.

**Comment by Ank4n** [comment]
> @Ank4n @kianenigma There is an issue which I noticed, StakingLedger.staking BoundedVec does not push new object (UnlockingChunks) instead it just increments the value inside UnlockingChunks. So the MaxUnlockingChunks bound in boundedVec is not quite useful, I think is a bug.

Can you elaborate more? May be refer in the code.

I see that when we unbond, a new UnlockChunk is pushed (or UnlockChunk for same era updated).
https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/mod.rs#L984

**Comment by MrishoLukamba** [comment]
> > @Ank4n @kianenigma There is an issue which I noticed, StakingLedger.staking BoundedVec does not push new object (UnlockingChunks) instead it just increments the value inside UnlockingChunks. So the MaxUnlockingChunks bound in boundedVec is not quite useful, I think is a bug.
> 
> Can you elaborate more? May be refer in the code.
> 
> I see that when we unbond, a new UnlockChunk is pushed (or UnlockChunk for same era updated). https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/mod.rs#L984

In the code it shows that,a new UnlockChunk is being pushed but in testing it is not pushed it is just updated the value field. So the bound MaxUnlockChunk placed there is not used at all

**Comment by Ank4n** [comment]
> > > @Ank4n @kianenigma There is an issue which I noticed, StakingLedger.staking BoundedVec does not push new object (UnlockingChunks) instead it just increments the value inside UnlockingChunks. So the MaxUnlockingChunks bound in boundedVec is not quite useful, I think is a bug.
> > 
> > 
> > Can you elaborate more? May be refer in the code.
> > I see that when we unbond, a new UnlockChunk is pushed (or UnlockChunk for same era updated). https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/mod.rs#L984
> 
> In the code it shows that,a new UnlockChunk is being pushed but in testing it is not pushed it is just updated the value field. So the bound MaxUnlockChunk placed there is not used at all. You can see in the new test i've added

@MrishoLukamba The behaviour looks correct to me. An `UnlockChunk` is only updated if multiple unbond happen in the same era. If you try to move forward in eras and `unbond`, you should see multiple chunks being added. If the `UnlockChunks` are already full, the transaction should return an error `NoMoreChunks` as the boundedVec is full.

@kianenigma Shouldn't we be able to keep the `MaxUnlockChunk` equal to `BondingDuration`. This will imply users can have `UnlockChunks` without claiming only for up to 28 eras. And in that case all the `UnlockChunks` are ready to be claimed (after 28 eras) so encouraging the user to just claim existing unlocks before requesting for more unlocks.
If you agree with above, we can bound `StakingLedger.unlocking` with `BondingDuration` and not have to introduce a new bound `MaxUnlockChunk`.

File: frame/staking/src/tests.rs [hunk 1]
```diff
@@ -1346,7 +1346,8 @@ fn too_many_unbond_calls_should_not_work() {
 	ExtBuilder::default().build_and_execute(|| {
 		let mut current_era = 0;
 		// locked at era MaxUnlockingChunks - 1 until 3
-		for i in 0..MaxUnlockingChunks::get() - 1 {
+		//maxunlockingchunks
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:1349]:
```suggestion
```

File: frame/staking/src/tests.rs [hunk 2]
```diff
@@ -1346,7 +1346,8 @@ fn too_many_unbond_calls_should_not_work() {
...
+		for i in 0..<<Test as Config>::MaxUnlockingChunks as Get::<u32>>::get()- 1 {
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:1350]:
```suggestion
		for i in 0..<<Test as Config>::MaxUnlockingChunks as Get<u32>>::get()-1 {
```

File: frame/staking/src/tests.rs [hunk 3]
```diff
@@ -1361,7 +1362,8 @@ fn too_many_unbond_calls_should_not_work() {
 		assert_ok!(Staking::unbond(RuntimeOrigin::signed(10), 1));
 		assert_eq!(
 			Staking::ledger(&10).unwrap().unlocking.len(),
-			MaxUnlockingChunks::get() as usize
+
+			<<Test as Config>::MaxUnlockingChunks as Get::<u32>>::get() as usize
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:1366]:
```suggestion
			<<Test as Config>::MaxUnlockingChunks as Get<u32>>::get() as usize
```

File: frame/staking/src/tests.rs [hunk 4]
```diff
@@ -5560,3 +5562,78 @@ fn reducing_history_depth_without_migration() {
 		HistoryDepth::set(original_history_depth);
 	});
 }
+
+#[test]
+fn change_max_unlocking_chunks_effect(){
+	//Concern is on validators only
+	//By Default 11, 10 are stash and ctrl and 21,20
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5569]:
```suggestion
	// Concern is on validators only
	// By Default 11, 10 are stash and ctrl and 21,20
```

Comment (kianenigma) [reply frame/staking/src/tests.rs:5569]: Please fix all comments accordingly.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Logic seems correct, but code needs to be formatted.

**Comment by MrishoLukamba** [comment]
> Logic seems correct, but code needs to be formatted.

Done, and should i use cargo fmt?

**Comment by Ank4n** [comment]
@MrishoLukamba I will close this PR and work on top of your code to get it merged. Pls add your Polkadot address in the PR for the tip.

Super thanks for your contribution. Feel free to follow the new PR here: https://github.com/paritytech/substrate/pull/12343

**Comment by MrishoLukamba** [comment]
> @MrishoLukamba I will close this PR and work on top of your code to get it merged. Pls add your Polkadot address in the PR for the tip.
> 
> Super thanks for your contribution. Feel free to follow the new PR here: #12343

Thanks, will also continue contributing on other issues.

**Comment by MrishoLukamba** [comment]
> /tip small

Sorry @kianenigma @Ank4n https://polkadot.js.org/apps/#/treasury/tips, the tip didnt go through

**Pull request closed without merging**